### DISCUSSION TITLE
FasTeX and failing activation due to line breaks

### DISCUSSION THREAD

**Discussion by tozGitHub** [WinEdt-Team/WinEdt-Support#11] [1.3 Ideas and features requests]
Dear Alex, dear community,

There is a slight incompatibility between WinEdt's wrapping mechanism and the use of the FasTeX plugin.

The plugin is basically a file of definitions of active strings. The standard is that each such string ends with two spaces (in order not to be activated by accident). However, if we are at the end of a line, then sometimes the activation fails (I think as we jump on a new line while typing the spaces).

It would be nice if line breaks would not prevent the active strings from being activated.

Best,
Thomas

More information about FasTeX for the curious among you:
The link to WinEdt.org: https://www.winedt.org/macros/latex/FasTeX.html
The TUGboat article about it: https://tug.org/TUGboat/tb16-4/tb49mars.pdf
A corrigendum of the article: https://tug.org/TUGboat/tb17-1/tb50mars.pdf

**Comment by Alex-Simonic** [comment]
I could not reproduce any problems with active strings ending with two spaces. For me, they were always activated and WinEdt did not jump to a new line while I was typing trailing spaces at the end of the line. I did not test with the actual FasTeX definitions but the idea was the same.

The details may depend on a lot of things: your wrapping method, trimming spaces at the end of the line (by default disabled for TeX documents), fixed or variable right margin, etc...

Perhaps you should test with default settings and then proceed to find any custom document setting that may be causing problems for you.

**Comment by tozGitHub** [comment]
Dear Alex,

Thank you for your answer. I looked further into it and now also could reproduce it with basically a clean WinEdt. I did not realise at first that it does not always happen, and it took me some effort to cause it. One setting where it might occur is when a line is very long, but WinEdt cannot introduce a line break.

In the example below, I try to use
```
xa<space><space>
```
which should be replaced by
```
\alpha
```
(it works in some lines)
Look at line 6. It extends beyond the window, and WinEdt cannot introduce a line break. I especially arranged the size of the WinEdt window in order that it looks like this:
![grafik](https://github.com/WinEdt-Team/WinEdt-Support/assets/52357589/c65c7d55-ffc0-445b-b29d-466052cbbcca)
First, I add an x at the end of the visible part of the line:
![grafik](https://github.com/WinEdt-Team/WinEdt-Support/assets/52357589/d02c3e3b-c3fd-49a1-894e-96478711cf88)
Now an a:
![grafik](https://github.com/WinEdt-Team/WinEdt-Support/assets/52357589/317e8515-af88-4ec4-b415-30ad737f19c7)
A space, and then a soft line break occurs and the cursor jumps to the next line:
![grafik](https://github.com/WinEdt-Team/WinEdt-Support/assets/52357589/a4d6acae-cf39-4f67-a788-979b6fe6943d)
Further spaces do not change anything anymore.
The TeX-content is:
```
\begin{dmath*}
d=
\sum_{g\in H_m}
\sum_{k=0}^{7^N-1}\left(\varphi\left(\frac{x+k}{7^N}+\right)-\varphi\left(\frac{a+k}{7^N}-\right)\right)
=\sum_{l=0}^{7^{N(m-1)}-1}
\sum_{k=0}^{7^N-1}\left(\varphi\left(\frac{\frac{y_0+i}{7^{N(m-1)}}+k}{7^N}+\right)-\varphi\left(\frac{\frac{x_0+i}{7^{N(m-1)}}+k}{7^N}-\right)\right)
=\sum_{i=0}^{7^{N(m-1)}-1}
\sum_{p=0}^{7^N-1}\left(
\varphi\left(\frac{x_1+i+ 7^{N(m-1)}k}{7^{Nm}}+\right)-\varphi\left(\frac{x_9+i+ 7^{N(m-1)} k}{8^{Nm}}-\right)\right)
\end{dmath*} 
```
That is what I added to ActiveStrings.ini:
```
STRING="xa  "
  ENABLED=1
  MODE_FILTER="TeX"
  MACRO="[BeginGroup;Backspace(4);Ins('\alpha');EndGroup;]"  
```

WinEdt version: Build: 20230519  (v. 11.1) - 64-bit
Windows version: 10.0.22631 Build 22631.

Best,
Thomas

**Reply by Alex-Simonic** [reply]
I see: if you have a long spaceless blob of a line then inserting a space inside it triggers soft wrapping and the caret moves to the next line (which is a standard behavior). This explains why an active string ending with more than one space will not be triggered under such circumstances.

The explanation is good enough for me as I do not consider it an issue that merits any further attention. In other words, I won't do anything about it as any "fix" would likely result in some other oddity like this...

Even if WinEdt would allow active strings past soft line breaks it would result in annoying flickering as the line would get wrapped and unwrapped after the second space. Knowing this it might be a better idea to change the triggers to a single space or two commas that would not cause any wrapping issues. Unwanted activation can be canceled with a single Undo (as such activations are always a possibility with so many active string abbreviations).
